### ISSUE TITLE
Path Length Example not working in 2.8

### ISSUE THREAD

**Issue by morph3us-net** [nortikin/sverchok#2489]
## Problem statement

I am trying to reproduce the example given in the documents (with a slight twist) [at the the bottom of this page](http://nikitron.cc.ua/sverch/html/nodes/analyzers/path_length.html).

<img width="1873" alt="screengrab_sverchok002" src="https://user-images.githubusercontent.com/43741267/61177260-e4af8700-a5d0-11e9-85aa-74221fff552a.png">

[I uploaded the scene on my megaupload account](https://mega.nz/#!oxpA1aBb!91ATdDKczfPEEB3qnfJxAXgYQjq3kEzrFyOXP8ItF5c)

Two problems arise (I am really sorry, I dont know GitHub too well, is it okay to write on usage problems here, too, or is it "only" bug fixing related issues?):
1.) The sampled points are not on top of the picked curve.
2.) I would like the generated boxes to rotate around their z-axis to follow the curve. Is there any way to access  that data? (Similar to the "duplicate objects along edge" node, however that one does not let me control how far apart the cubes are)

My error log:
```
sv: import settings
sv: import all modules
[Errno 2] No such file or directory: 'C:\\Users\\morph3us\\AppData\\Roaming\\Blender Foundation\\Blender\\2.80\\datafiles\\sverchok\\sv_fx.wad'
seeing this message means you need to generate an uptoken
sv: append app.handlers:
sv: enable internal debug logging.
INFO:sverchok.utils.logging:162:log level, 20
 ______ _    _ _______  ______ _______ _     _  _____  _     _
/______  \  /  |______ |_____/ |       |_____| |     | |____/
______/   \/   |______ |    \_ |_____  |     | |_____| |    \_
initialized.

sv: version: 0.6.0.0
sv: 300 nodes.
Read blend: C:\Users\morph3us\.Nimi Places\Containers\Quick access\Projects\GoogleMaps\SverchokTest002.blend
INFO:root:140:Initializing Sverchok logging. Blender version 2.80 (sub 74), Sverchok version 0.6.0.0
ERROR:sverchok.core.update_system:162:Node Path Length had exception: No data passed into socket `Vertices'
Traceback (most recent call last):
  File "C:\Blender\blender-2.80rc1-windows64\2.80\scripts\addons_contrib\sverchok\core\update_system.py", line 325, in do_update_general
    node.process()
  File "C:\Blender\blender-2.80rc1-windows64\2.80\scripts\addons_contrib\sverchok\nodes\analyzer\path_length.py", line 117, in process
    meshes = self.get_data()
  File "C:\Blender\blender-2.80rc1-windows64\2.80\scripts\addons_contrib\sverchok\nodes\analyzer\path_length.py", line 96, in get_data
    vertices = si['Vertices'].sv_get()
  File "C:\Blender\blender-2.80rc1-windows64\2.80\scripts\addons_contrib\sverchok\sockets.py", line 356, in sv_get
    source_data = SvGetSocket(self, deepcopy = True if self.needs_data_conversion() else deepcopy)
  File "C:\Blender\blender-2.80rc1-windows64\2.80\scripts\addons_contrib\sverchok\core\socket_data.py", line 108, in SvGetSocket
    raise SvNoDataError(socket)
sverchok.core.socket_data.SvNoDataError: No data passed into socket `Vertices'
```

## Expected result

The boxes should be on top of the curve, the origin of the generated object is (0, 0, 0), whereas my curve's is located at (0.860726m, 4.06574m, -0m). 
Moving the curve's origin to (0, 0, 0) fixes the issue and the boxes are now aligned on top of the curve. So maybe the node tree needs to add the curves location somewhere? I mean, putting all your curve's origins constantly to the scene origin is kinda cumbersome, no?

Still, here's the issue with the rotation, though, which is frankly not a bug, but maybe someone can help, please?

edit:
I used [this](https://cloud.githubusercontent.com/assets/284644/6096654/b300fbfa-afbf-11e4-901b-1361a44238c2.png) as a starting point for the rotational stuff. I am now able to rotate each cube individually randomly. I now only need to make them follow the path...

## Sverchok version
"b28_prelease_master": 0.072
installed via #2326

**Comment by morph3us-net** [comment]
I thought that maybe I could get the "Vector Interpolations'" Vertices output and convert their normals direction to a degree with the following node setup. As you can see, however, the stethoscope outputs radians, I believe? I may be totally wrong and off, though... Could someone hint me in the right direction, please? :)

<img width="1871" alt="screengrab_sverchok003" src="https://user-images.githubusercontent.com/43741267/61183434-b6b65b00-a641-11e9-81ef-6a70aa71c828.png">

**Comment by zeffii** [comment]
1.  Sampled points (objects in) are not automatically transformed by the Matrix of the picked Curve Object, you must either 
- apply the matrix / transforms to the Object before picking up the object
- or, apply the matrix to the vertices using `Matrix Apply` node

2.   @portnov  that's up your street to answer :)

**Comment by morph3us-net** [comment]
Thank you very much for the clarification! I feel a little overwhelmed by all of this, so please forgive me if I thought this is a bug... I have the matrix issue fixed now!

**Comment by portnov** [comment]
Well, you are converting Z output of "vector polar out" to stetoscope — that's Z coordinate, I think it really is zero. Try phi output instead.

**Comment by zeffii** [comment]
@morph3us-net  you're welcome.  There's a lot to take in with Sverchok, it is truly a massive add-on, sometimes we make user-faced-ui-decisions that may not be intuitive.  other times....it's complicated for a reason :)

**Comment by morph3us-net** [comment]
I have tried phi, too, unfortunately it does not behave like the expected outcome...

<img width="1877" alt="screengrab_sverchok004" src="https://user-images.githubusercontent.com/43741267/61184698-37c91e80-a651-11e9-810c-a5e70b90e015.png">

It totally makes sense to put that into its own node, I get it now, it was a good decision, only.... unexpected :P

**Comment by portnov** [comment]
It seems you have to rotate all normals 90 degrees. Try to add (or substract) 90 to phi output by use of Math node.

**Comment by portnov** [comment]
Or, you can just rotate you bar object (don't forget to apply rotation by Ctrl-A) :)

**Comment by portnov** [comment]
Ah, and "Calculate normals" probably calculates them incorrectly if Edges input is not connected. Try connect it.

**Comment by morph3us-net** [comment]
Its not a 90° problem, as not all of the boxes are facing wrongly in the exact same rotation.
Does the Math Node iterate over the matrix list? It gets red immediatly here.

Connecting the edges does neither change the stethoscope output, not the rotations...

**Comment by portnov** [comment]
It seems Blender understands "vertex normals" a bit differently than you and I :)
Try the following:
* convert your curve into mesh (Alt-C)
* go to edit mode and enable display of vertex normals (in 2.79 it's N panel -> Mesh display -> Normals, should be similar in 2.8)

Then you will see what Blender means by "vertex normal" for a curve...

**Comment by portnov** [comment]
![Screenshot_20190714_233745](https://user-images.githubusercontent.com/284644/61187859-661a1e80-a690-11e9-83cc-b9355aa0a2eb.png)

**Comment by Durman** [comment]
From point of view of Blender normal of point is vector first point of which is (0, 0, 0) and second is coordinates of the point.

**Comment by morph3us-net** [comment]
Hmm... Seems I am too dumb to figure this one out, then...
I mean, if all of these normals are fanning out from the scene's origin, it doesnt help me at all.

Maybe I could get the points before and after and then interpolate between their positions? 
Sounds pretty computation heavy, though, if I would like to do that with a longer curve...

**Comment by Durman** [comment]
You are thinking in right direction. Such algorithm will take O(n) time where n us number of points of curve. It means that process time increasing linearly with increasing number of inputs points. It is pretty fast. Have a look this: [Computational complexity](https://en.wikipedia.org/wiki/Computational_complexity). 

And this calculation can be done with existing nodes of course. It is not very comfortable so there was thoughts about node which could calculate such normals. Unfortunately it remains as project yet #2242. But I'm thinking time to times to finish this node.))

**Comment by morph3us-net** [comment]
It would be really awesome if you could find the time to do that, as I am doing this for work and am on a time schedule. I would love to solve this puzzle on my own, but I fear I do not have the time to do so... :(

**Comment by Durman** [comment]
This node should be definitely done but it has low priority in my queue at this moment. My head is filled with edge intersection algorithm.))

**Comment by zeffii** [comment]
the duplicate_objects_along_edges node might suffice? 
![image](https://user-images.githubusercontent.com/619340/61201195-136a5200-a6e4-11e9-825e-252190e8268e.png)

**Comment by morph3us-net** [comment]
unfortunately that node does not let me space out the boxes evenly...
I would like to tell it to have a box, for example, every 1.5 meters.

**Comment by zeffii** [comment]
Evenly spacing is a separate issue. The spacing is dictated by the distance between verts+edges derived from the input Curve.

![image](https://user-images.githubusercontent.com/619340/61203987-4284c180-a6ec-11e9-9c1b-92bf38c5249a.png)

you would stick the path length node inbetween to find the total length of the path and a math + float range node too, to obtain the `[0., .....   1]`  range for an evenly subdivided path

![image](https://user-images.githubusercontent.com/619340/61204639-0a7e7e00-a6ee-11e9-96e5-6b3924d6a47c.png)

**Comment by morph3us-net** [comment]
Your proposal makes sense! I had tried this node before #2484 but had some issues. I cannot recall which ones exactly, but I will try it out again as soon as possible, as I now know more about sverchok in general.

I believe that at some point, if I made the spacing too far apart, the setup would put multiple copies of a box in between two vertices (e.g. the result was as expected when having a distance of 50 centimeter, but not 3 meters). But this is from the top of my head, right now I have to work on other stuff in the scene...

**Comment by zeffii** [comment]
@morph3us-net hang on in there :)

**Comment by morph3us-net** [comment]
I got it working, however the issue arises that I mentioned.
I am done with work now and will try more when I am home in 2 hours! :)

Thank you guys for having an open ear, you everybody`s so helpfull, I am really gratefull for that!
![sverchok_grouping](https://user-images.githubusercontent.com/43741267/61227657-00c43d00-a725-11e9-89a1-cf85ddc05874.jpg)
[And here`s the scene](https://mega.nz/#!ApAylKBS!p-JSHp3OXkLxYt5RUbT_ZLLtqWMAdL45iiKsWQMckaw)

**Comment by morph3us-net** [comment]
I was thinking of maybe using the "scale: count" mode and somehow compute the padding, when I noticed the problem I marked with the red arrow. I am going back to the direct matrix manipulation, lets see if I can get these vectors somehow...

<img width="1875" alt="screengrab_sverchok006" src="https://user-images.githubusercontent.com/43741267/61245317-7f80a080-a74c-11e9-8607-64219435140c.png">

**Comment by zeffii** [comment]
i would "resample" high,  using the tecnique i show above, to get a nicely subdivided vertex sequence. Then from that sequence every n-th vertex.  You'll likely get verts that very closely match the original bezier path

**Comment by morph3us-net** [comment]
Sounds good, so I need to somehow manipulate the edgesR.1 input of the "duplicate along edges" node?
I suspect one of the "list" mnodes? My head hurts...

**Comment by zeffii** [comment]
pass an IntRange node into a List Item node to pick every nth

**Comment by morph3us-net** [comment]
zeffii, you cant imagine how grateful and happy I am!
This is not a perfect solution, but it works for now!
Thank you very much for the time you invested!

<img width="1877" alt="screengrab_sverchok007" src="https://user-images.githubusercontent.com/43741267/61250605-6382fc00-a758-11e9-9a3d-5f9e058749ce.png">

**Comment by zeffii** [comment]
the more your input curve resembles a plausible train-track-path, the better these kinds of things work.   Ie, there's a minimum radius that a curve/corner can have for carriages, the example you show has a sudden curve change with a radius much smaller than the sleepers would allow in real life too.

**Comment by zeffii** [comment]
but if you wanted the sleepers to be placed directly on the verts and not the middle of the edges, then ...yeah.. i agree there should perhaps be a simpler way.

**Comment by morph3us-net** [comment]
![screenshot_20190716_120358](https://user-images.githubusercontent.com/43741267/61285832-e9875d00-a7c1-11e9-86bc-a9eb40a99b8b.jpg)

I am amazed on how much can go wrong... I have been trying to sweep the actual tracks for an hour and got nowhere. I imported the node tree from one scene to the other and am getting wierd errors with "object xyz is not in the view layer" (which i never reference anywhere in the tree).

I got it to work with attaching the other curves to the original one, rather than picking the new ones.
However, as there are several pathes in the curve, it scatters in between the last point of one curve to the first part of the next seperate.
I am really near to giving up! Too bad :(
